File: PepesComing/Controller.vb
Imports Microsoft.Xna.Framework
Imports Microsoft.Xna.Framework.Graphics
Imports Microsoft.Xna.Framework.Input

Public Class Controller
    Private previousKeyboardState As KeyboardState
    Private keyboardState As KeyboardState
    Private mouseState As MouseState
    Private previousMouseState As MouseState

    Private DragStart As Vector2

    Private Const CAMERA_KEY_UP As Keys = Keys.W
    Private Const CAMERA_KEY_DOWN As Keys = Keys.S
    Private Const CAMERA_KEY_LEFT As Keys = Keys.A
    Private Const CAMERA_KEY_RIGHT As Keys = Keys.D
    Private Const CAMERA_KEY_ZOOM_IN As Keys = Keys.OemPlus
    Private Const CAMERA_KEY_ZOOM_OUT As Keys = Keys.OemMinus

    Private Const REGENERATE_MAZE As Keys = Keys.R
    Private Const SOLVE_MAZE As Keys = Keys.P

    ReadOnly Property CameraUp As Boolean
        Get
            Return keyboardState.IsKeyDown(CAMERA_KEY_UP)
        End Get
    End Property

    ReadOnly Property CameraDown As Boolean
        Get
            Return keyboardState.IsKeyDown(CAMERA_KEY_DOWN)
        End Get
    End Property

    ReadOnly Property CameraLeft As Boolean
        Get
            Return keyboardState.IsKeyDown(CAMERA_KEY_LEFT)
        End Get
    End Property

    ReadOnly Property CameraRight As Boolean
        Get
            Return keyboardState.IsKeyDown(CAMERA_KEY_RIGHT)
        End Get
    End Property

    ReadOnly Property CameraZoomIn As Boolean
        Get
            Return keyboardState.IsKeyDown(CAMERA_KEY_ZOOM_IN)
        End Get
    End Property

    ReadOnly Property CameraZoomOut As Boolean
        Get
            Return keyboardState.IsKeyDown(CAMERA_KEY_ZOOM_OUT)
        End Get
    End Property

    ReadOnly Property Solve As Boolean
        Get
            Return keyboardState.IsKeyDown(SOLVE_MAZE) And Not previousKeyboardState.IsKeyDown(SOLVE_MAZE)
        End Get
    End Property

    ReadOnly Property RegenerateMaze As Boolean
        Get
            Return keyboardState.IsKeyDown(REGENERATE_MAZE) And Not previousKeyboardState.IsKeyDown(REGENERATE_MAZE)
        End Get
    End Property

    ReadOnly Property MouseDown As Boolean
        Get
            Return mouseState.LeftButton
        End Get
    End Property

    ReadOnly Property MousePosition As Vector2
        Get
            Return mouseState.Position.ToVector2
        End Get
    End Property

    Public Sub Update(keyboardState As KeyboardState, mouseState As MouseState)
        'Console.WriteLine("x: {0}, y: {1}, scroll {2}", mouseState.X, mouseState.Y, mouseState.ScrollWheelValue)
        Me.previousKeyboardState = Me.keyboardState
        Me.keyboardState = keyboardState
        Me.previousMouseState = Me.mouseState
        Me.mouseState = mouseState
    End Sub

    Public Sub DragCamera(camera As Camera)
        If mouseState.LeftButton = ButtonState.Pressed Then
            If previousMouseState.LeftButton = ButtonState.Released Then
                DragStart = MousePosition
                camera.DragCamera(DragStart, MousePosition)
            End If
            camera.Dragging(MousePosition)
        End If
    End Sub
End Class

File: PepesComing.ConvertedToC#/Controller.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;
using System.Threading.Tasks;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
namespace PepesComing
{

	public class Controller
	{
		private KeyboardState previousKeyboardState;
		private KeyboardState keyboardState;
		private MouseState mouseState;

		private MouseState previousMouseState;

		private Vector2 DragStart;
		private const Keys CAMERA_KEY_UP = Keys.W;
		private const Keys CAMERA_KEY_DOWN = Keys.S;
		private const Keys CAMERA_KEY_LEFT = Keys.A;
		private const Keys CAMERA_KEY_RIGHT = Keys.D;
		private const Keys CAMERA_KEY_ZOOM_IN = Keys.OemPlus;

		private const Keys CAMERA_KEY_ZOOM_OUT = Keys.OemMinus;
		private const Keys REGENERATE_MAZE = Keys.R;

		private const Keys SOLVE_MAZE = Keys.P;
		public bool CameraUp {
			get { return keyboardState.IsKeyDown(CAMERA_KEY_UP); }
		}

		public bool CameraDown {
			get { return keyboardState.IsKeyDown(CAMERA_KEY_DOWN); }
		}

		public bool CameraLeft {
			get { return keyboardState.IsKeyDown(CAMERA_KEY_LEFT); }
		}

		public bool CameraRight {
			get { return keyboardState.IsKeyDown(CAMERA_KEY_RIGHT); }
		}

		public bool CameraZoomIn {
			get { return keyboardState.IsKeyDown(CAMERA_KEY_ZOOM_IN); }
		}

		public bool CameraZoomOut {
			get { return keyboardState.IsKeyDown(CAMERA_KEY_ZOOM_OUT); }
		}

		public bool Solve {
			get { return keyboardState.IsKeyDown(SOLVE_MAZE) & !previousKeyboardState.IsKeyDown(SOLVE_MAZE); }
		}

		public bool RegenerateMaze {
			get { return keyboardState.IsKeyDown(REGENERATE_MAZE) & !previousKeyboardState.IsKeyDown(REGENERATE_MAZE); }
		}

		public bool MouseDown {
			get { return mouseState.LeftButton == ButtonState.Pressed; }
		}

		public Vector2 MousePosition {
			get { return mouseState.Position.ToVector2(); }
		}

		public void Update(KeyboardState keyboardState, MouseState mouseState)
		{
			//Console.WriteLine("x: {0}, y: {1}, scroll {2}", mouseState.X, mouseState.Y, mouseState.ScrollWheelValue)
			this.previousKeyboardState = this.keyboardState;
			this.keyboardState = keyboardState;
			this.previousMouseState = this.mouseState;
			this.mouseState = mouseState;
		}

		public void DragCamera(Camera camera)
		{
			if (mouseState.LeftButton == ButtonState.Pressed) {
				if (previousMouseState.LeftButton == ButtonState.Released) {
					DragStart = MousePosition;
					camera.DragCamera(DragStart, MousePosition);
				}
				camera.Dragging(MousePosition);
			}
		}
	}
}
